openedx/core/djangoapps/user_api/models.py

```diff
@@ -25,28 +56,79 @@ class Meta:  # pylint: disable=missing-docstring
         unique_together = ("user", "key")
 
     @classmethod
-    def set_preference(cls, user, preference_key, preference_value):
+    def set_preference(cls, user, preference_key, preference_value, save=True):
+        """Sets the user preference for a given key, creating it if it doesn't exist.
+
+        Arguments:
+            user (User): The user whose preference should be set.
+            preference_key (string): The key for the user preference.
+            preference_value (string): The value to be stored.
+            save (boolean): If true then save the model (defaults to True).
+
+        Raises:
+            ValidationError: the update was rejected because it was invalid
         """
-        Sets the user preference for a given key
+        user_preference, _ = cls.objects.get_or_create(user=user, key=preference_key)
+        user_preference.value = preference_value
+        user_preference.full_clean()
+        if save:
+            user_preference.save()
+
+    @classmethod
+    def validate_preference(cls, user, preference_key, preference_value):
+        """Validates the combination of preference key and value.
+
+        Arguments:
+            user (User): The user whose preference should be validated.
+            preference_key (string): The key for the user preference.
+            preference_value (string): The value to be stored.
+
+        Raises:
+            ValidationError: returned if the key and/or value is invalid.
         """
-        user_pref, _ = cls.objects.get_or_create(user=user, key=preference_key)
-        user_pref.value = preference_value
-        user_pref.save()
+        user_preference, _ = cls.objects.get_or_create(user=user, key=preference_key)
+        old_value = user_preference.value
+        user_preference.value = preference_value
+        try:
+            user_preference.full_clean()
```

Comment:
This is the Django model way of doing validation. I'm going to take your suggestion of using a serializer instead to do the validation.